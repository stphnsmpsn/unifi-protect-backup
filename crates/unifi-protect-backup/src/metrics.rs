use crate::{
    archive::borg::Metrics as BorgArchiveMetrics,
    backup::{local::Metrics as LocalBackupMetrics, rclone::Metrics as RcloneBackupMetrics},
};
use hyper::{Request, Response, body::Incoming, server::conn::http1, service::service_fn};
use hyper_util::rt::TokioIo;
use serde::Serialize;
use std::{net::SocketAddr, sync::Arc};
use tokio::net::TcpListener;

#[derive(Default, Serialize)]
pub struct Metrics {
    pub local_backup: Arc<LocalBackupMetrics>,
    pub rclone_backup: Arc<RcloneBackupMetrics>,
    pub borg_archive: Arc<BorgArchiveMetrics>,
}

pub async fn start_metrics_server(
    metrics: Arc<Metrics>,
    address: &str,
    port: u16,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let addr: SocketAddr = format!("{address}:{port}").parse()?;
    let listener = TcpListener::bind(addr).await?;

    tracing::info!("Metrics server listening on http://{addr}");

    loop {
        let (stream, _) = listener.accept().await?;
        let io = TokioIo::new(stream);
        let metrics = metrics.clone();

        tokio::task::spawn(async move {
            if let Err(err) = http1::Builder::new()
                .serve_connection(io, service_fn(|req| handle_request(req, metrics.clone())))
                .await
            {
                tracing::error!("Error serving connection: {:?}", err);
            }
        });
    }
}

async fn handle_request(
    req: Request<Incoming>,
    metrics: Arc<Metrics>,
) -> Result<Response<String>, hyper::Error> {
    match req.uri().path() {
        "/metrics" => {
            let prometheus_output =
                serde_prometheus::to_string(&*metrics, None, std::collections::HashMap::new())
                    .unwrap_or_else(|e| format!("Error serializing metrics: {e}"));

            Ok(Response::builder()
                .status(200)
                .header("Content-Type", "text/plain; version=0.0.4; charset=utf-8")
                .body(prometheus_output)
                .unwrap())
        }
        _ => Ok(Response::builder()
            .status(404)
            .body("Not Found".to_string())
            .unwrap()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    pub fn test_metrics() {
        insta::assert_snapshot!(
            serde_prometheus::to_string(
                &Metrics::default(),
                None,
                std::collections::HashMap::new()
            )
            .unwrap()
        );
    }
}
