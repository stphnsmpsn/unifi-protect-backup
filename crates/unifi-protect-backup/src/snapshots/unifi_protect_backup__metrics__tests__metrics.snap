---
source: crates/unifi-protect-backup/src/metrics.rs
expression: "serde_prometheus::to_string(&Metrics::default(), None,\nstd::collections::HashMap::new()).unwrap()"
---
hit_count{path = "local_backup/prune_directory"} 0
throughput_samples{path = "local_backup/prune_directory"} 0
throughput_min{path = "local_backup/prune_directory"} 0
throughput_max{path = "local_backup/prune_directory"} 0
throughput_mean{path = "local_backup/prune_directory"} 0
throughput_stdev{path = "local_backup/prune_directory"} 0
throughput{quantile = "0.9", path = "local_backup/prune_directory"} 0
throughput{quantile = "0.95", path = "local_backup/prune_directory"} 0
throughput{quantile = "0.99", path = "local_backup/prune_directory"} 0
throughput{quantile = "0.999", path = "local_backup/prune_directory"} 0
throughput{quantile = "0.9999", path = "local_backup/prune_directory"} 0
error_count{path = "local_backup/prune_directory"} 0
response_time_samples{path = "local_backup/prune_directory"} 0
response_time_min{path = "local_backup/prune_directory"} 0
response_time_max{path = "local_backup/prune_directory"} 0
response_time_mean{path = "local_backup/prune_directory"} 0
response_time_stdev{path = "local_backup/prune_directory"} 0
response_time{quantile = "0.9", path = "local_backup/prune_directory"} 0
response_time{quantile = "0.95", path = "local_backup/prune_directory"} 0
response_time{quantile = "0.99", path = "local_backup/prune_directory"} 0
response_time{quantile = "0.999", path = "local_backup/prune_directory"} 0
response_time{quantile = "0.9999", path = "local_backup/prune_directory"} 0
hit_count{path = "local_backup/backup"} 0
throughput_samples{path = "local_backup/backup"} 0
throughput_min{path = "local_backup/backup"} 0
throughput_max{path = "local_backup/backup"} 0
throughput_mean{path = "local_backup/backup"} 0
throughput_stdev{path = "local_backup/backup"} 0
throughput{quantile = "0.9", path = "local_backup/backup"} 0
throughput{quantile = "0.95", path = "local_backup/backup"} 0
throughput{quantile = "0.99", path = "local_backup/backup"} 0
throughput{quantile = "0.999", path = "local_backup/backup"} 0
throughput{quantile = "0.9999", path = "local_backup/backup"} 0
error_count{path = "local_backup/backup"} 0
response_time_samples{path = "local_backup/backup"} 0
response_time_min{path = "local_backup/backup"} 0
response_time_max{path = "local_backup/backup"} 0
response_time_mean{path = "local_backup/backup"} 0
response_time_stdev{path = "local_backup/backup"} 0
response_time{quantile = "0.9", path = "local_backup/backup"} 0
response_time{quantile = "0.95", path = "local_backup/backup"} 0
response_time{quantile = "0.99", path = "local_backup/backup"} 0
response_time{quantile = "0.999", path = "local_backup/backup"} 0
response_time{quantile = "0.9999", path = "local_backup/backup"} 0
hit_count{path = "local_backup/prune"} 0
throughput_samples{path = "local_backup/prune"} 0
throughput_min{path = "local_backup/prune"} 0
throughput_max{path = "local_backup/prune"} 0
throughput_mean{path = "local_backup/prune"} 0
throughput_stdev{path = "local_backup/prune"} 0
throughput{quantile = "0.9", path = "local_backup/prune"} 0
throughput{quantile = "0.95", path = "local_backup/prune"} 0
throughput{quantile = "0.99", path = "local_backup/prune"} 0
throughput{quantile = "0.999", path = "local_backup/prune"} 0
throughput{quantile = "0.9999", path = "local_backup/prune"} 0
error_count{path = "local_backup/prune"} 0
response_time_samples{path = "local_backup/prune"} 0
response_time_min{path = "local_backup/prune"} 0
response_time_max{path = "local_backup/prune"} 0
response_time_mean{path = "local_backup/prune"} 0
response_time_stdev{path = "local_backup/prune"} 0
response_time{quantile = "0.9", path = "local_backup/prune"} 0
response_time{quantile = "0.95", path = "local_backup/prune"} 0
response_time{quantile = "0.99", path = "local_backup/prune"} 0
response_time{quantile = "0.999", path = "local_backup/prune"} 0
response_time{quantile = "0.9999", path = "local_backup/prune"} 0
hit_count{path = "rclone_backup/backup"} 0
throughput_samples{path = "rclone_backup/backup"} 0
throughput_min{path = "rclone_backup/backup"} 0
throughput_max{path = "rclone_backup/backup"} 0
throughput_mean{path = "rclone_backup/backup"} 0
throughput_stdev{path = "rclone_backup/backup"} 0
throughput{quantile = "0.9", path = "rclone_backup/backup"} 0
throughput{quantile = "0.95", path = "rclone_backup/backup"} 0
throughput{quantile = "0.99", path = "rclone_backup/backup"} 0
throughput{quantile = "0.999", path = "rclone_backup/backup"} 0
throughput{quantile = "0.9999", path = "rclone_backup/backup"} 0
error_count{path = "rclone_backup/backup"} 0
response_time_samples{path = "rclone_backup/backup"} 0
response_time_min{path = "rclone_backup/backup"} 0
response_time_max{path = "rclone_backup/backup"} 0
response_time_mean{path = "rclone_backup/backup"} 0
response_time_stdev{path = "rclone_backup/backup"} 0
response_time{quantile = "0.9", path = "rclone_backup/backup"} 0
response_time{quantile = "0.95", path = "rclone_backup/backup"} 0
response_time{quantile = "0.99", path = "rclone_backup/backup"} 0
response_time{quantile = "0.999", path = "rclone_backup/backup"} 0
response_time{quantile = "0.9999", path = "rclone_backup/backup"} 0
hit_count{path = "rclone_backup/prune"} 0
throughput_samples{path = "rclone_backup/prune"} 0
throughput_min{path = "rclone_backup/prune"} 0
throughput_max{path = "rclone_backup/prune"} 0
throughput_mean{path = "rclone_backup/prune"} 0
throughput_stdev{path = "rclone_backup/prune"} 0
throughput{quantile = "0.9", path = "rclone_backup/prune"} 0
throughput{quantile = "0.95", path = "rclone_backup/prune"} 0
throughput{quantile = "0.99", path = "rclone_backup/prune"} 0
throughput{quantile = "0.999", path = "rclone_backup/prune"} 0
throughput{quantile = "0.9999", path = "rclone_backup/prune"} 0
error_count{path = "rclone_backup/prune"} 0
response_time_samples{path = "rclone_backup/prune"} 0
response_time_min{path = "rclone_backup/prune"} 0
response_time_max{path = "rclone_backup/prune"} 0
response_time_mean{path = "rclone_backup/prune"} 0
response_time_stdev{path = "rclone_backup/prune"} 0
response_time{quantile = "0.9", path = "rclone_backup/prune"} 0
response_time{quantile = "0.95", path = "rclone_backup/prune"} 0
response_time{quantile = "0.99", path = "rclone_backup/prune"} 0
response_time{quantile = "0.999", path = "rclone_backup/prune"} 0
response_time{quantile = "0.9999", path = "rclone_backup/prune"} 0
hit_count{path = "rclone_backup/single_stream_upload"} 0
throughput_samples{path = "rclone_backup/single_stream_upload"} 0
throughput_min{path = "rclone_backup/single_stream_upload"} 0
throughput_max{path = "rclone_backup/single_stream_upload"} 0
throughput_mean{path = "rclone_backup/single_stream_upload"} 0
throughput_stdev{path = "rclone_backup/single_stream_upload"} 0
throughput{quantile = "0.9", path = "rclone_backup/single_stream_upload"} 0
throughput{quantile = "0.95", path = "rclone_backup/single_stream_upload"} 0
throughput{quantile = "0.99", path = "rclone_backup/single_stream_upload"} 0
throughput{quantile = "0.999", path = "rclone_backup/single_stream_upload"} 0
throughput{quantile = "0.9999", path = "rclone_backup/single_stream_upload"} 0
error_count{path = "rclone_backup/single_stream_upload"} 0
response_time_samples{path = "rclone_backup/single_stream_upload"} 0
response_time_min{path = "rclone_backup/single_stream_upload"} 0
response_time_max{path = "rclone_backup/single_stream_upload"} 0
response_time_mean{path = "rclone_backup/single_stream_upload"} 0
response_time_stdev{path = "rclone_backup/single_stream_upload"} 0
response_time{quantile = "0.9", path = "rclone_backup/single_stream_upload"} 0
response_time{quantile = "0.95", path = "rclone_backup/single_stream_upload"} 0
response_time{quantile = "0.99", path = "rclone_backup/single_stream_upload"} 0
response_time{quantile = "0.999", path = "rclone_backup/single_stream_upload"} 0
response_time{quantile = "0.9999", path = "rclone_backup/single_stream_upload"} 0
hit_count{path = "rclone_backup/chunked_stream_upload"} 0
throughput_samples{path = "rclone_backup/chunked_stream_upload"} 0
throughput_min{path = "rclone_backup/chunked_stream_upload"} 0
throughput_max{path = "rclone_backup/chunked_stream_upload"} 0
throughput_mean{path = "rclone_backup/chunked_stream_upload"} 0
throughput_stdev{path = "rclone_backup/chunked_stream_upload"} 0
throughput{quantile = "0.9", path = "rclone_backup/chunked_stream_upload"} 0
throughput{quantile = "0.95", path = "rclone_backup/chunked_stream_upload"} 0
throughput{quantile = "0.99", path = "rclone_backup/chunked_stream_upload"} 0
throughput{quantile = "0.999", path = "rclone_backup/chunked_stream_upload"} 0
throughput{quantile = "0.9999", path = "rclone_backup/chunked_stream_upload"} 0
error_count{path = "rclone_backup/chunked_stream_upload"} 0
response_time_samples{path = "rclone_backup/chunked_stream_upload"} 0
response_time_min{path = "rclone_backup/chunked_stream_upload"} 0
response_time_max{path = "rclone_backup/chunked_stream_upload"} 0
response_time_mean{path = "rclone_backup/chunked_stream_upload"} 0
response_time_stdev{path = "rclone_backup/chunked_stream_upload"} 0
response_time{quantile = "0.9", path = "rclone_backup/chunked_stream_upload"} 0
response_time{quantile = "0.95", path = "rclone_backup/chunked_stream_upload"} 0
response_time{quantile = "0.99", path = "rclone_backup/chunked_stream_upload"} 0
response_time{quantile = "0.999", path = "rclone_backup/chunked_stream_upload"} 0
response_time{quantile = "0.9999", path = "rclone_backup/chunked_stream_upload"} 0
hit_count{path = "rclone_backup/temp_file_upload"} 0
throughput_samples{path = "rclone_backup/temp_file_upload"} 0
throughput_min{path = "rclone_backup/temp_file_upload"} 0
throughput_max{path = "rclone_backup/temp_file_upload"} 0
throughput_mean{path = "rclone_backup/temp_file_upload"} 0
throughput_stdev{path = "rclone_backup/temp_file_upload"} 0
throughput{quantile = "0.9", path = "rclone_backup/temp_file_upload"} 0
throughput{quantile = "0.95", path = "rclone_backup/temp_file_upload"} 0
throughput{quantile = "0.99", path = "rclone_backup/temp_file_upload"} 0
throughput{quantile = "0.999", path = "rclone_backup/temp_file_upload"} 0
throughput{quantile = "0.9999", path = "rclone_backup/temp_file_upload"} 0
error_count{path = "rclone_backup/temp_file_upload"} 0
response_time_samples{path = "rclone_backup/temp_file_upload"} 0
response_time_min{path = "rclone_backup/temp_file_upload"} 0
response_time_max{path = "rclone_backup/temp_file_upload"} 0
response_time_mean{path = "rclone_backup/temp_file_upload"} 0
response_time_stdev{path = "rclone_backup/temp_file_upload"} 0
response_time{quantile = "0.9", path = "rclone_backup/temp_file_upload"} 0
response_time{quantile = "0.95", path = "rclone_backup/temp_file_upload"} 0
response_time{quantile = "0.99", path = "rclone_backup/temp_file_upload"} 0
response_time{quantile = "0.999", path = "rclone_backup/temp_file_upload"} 0
response_time{quantile = "0.9999", path = "rclone_backup/temp_file_upload"} 0
hit_count{path = "borg_archive/archive"} 0
throughput_samples{path = "borg_archive/archive"} 0
throughput_min{path = "borg_archive/archive"} 0
throughput_max{path = "borg_archive/archive"} 0
throughput_mean{path = "borg_archive/archive"} 0
throughput_stdev{path = "borg_archive/archive"} 0
throughput{quantile = "0.9", path = "borg_archive/archive"} 0
throughput{quantile = "0.95", path = "borg_archive/archive"} 0
throughput{quantile = "0.99", path = "borg_archive/archive"} 0
throughput{quantile = "0.999", path = "borg_archive/archive"} 0
throughput{quantile = "0.9999", path = "borg_archive/archive"} 0
error_count{path = "borg_archive/archive"} 0
response_time_samples{path = "borg_archive/archive"} 0
response_time_min{path = "borg_archive/archive"} 0
response_time_max{path = "borg_archive/archive"} 0
response_time_mean{path = "borg_archive/archive"} 0
response_time_stdev{path = "borg_archive/archive"} 0
response_time{quantile = "0.9", path = "borg_archive/archive"} 0
response_time{quantile = "0.95", path = "borg_archive/archive"} 0
response_time{quantile = "0.99", path = "borg_archive/archive"} 0
response_time{quantile = "0.999", path = "borg_archive/archive"} 0
response_time{quantile = "0.9999", path = "borg_archive/archive"} 0
hit_count{path = "borg_archive/prune"} 0
throughput_samples{path = "borg_archive/prune"} 0
throughput_min{path = "borg_archive/prune"} 0
throughput_max{path = "borg_archive/prune"} 0
throughput_mean{path = "borg_archive/prune"} 0
throughput_stdev{path = "borg_archive/prune"} 0
throughput{quantile = "0.9", path = "borg_archive/prune"} 0
throughput{quantile = "0.95", path = "borg_archive/prune"} 0
throughput{quantile = "0.99", path = "borg_archive/prune"} 0
throughput{quantile = "0.999", path = "borg_archive/prune"} 0
throughput{quantile = "0.9999", path = "borg_archive/prune"} 0
error_count{path = "borg_archive/prune"} 0
response_time_samples{path = "borg_archive/prune"} 0
response_time_min{path = "borg_archive/prune"} 0
response_time_max{path = "borg_archive/prune"} 0
response_time_mean{path = "borg_archive/prune"} 0
response_time_stdev{path = "borg_archive/prune"} 0
response_time{quantile = "0.9", path = "borg_archive/prune"} 0
response_time{quantile = "0.95", path = "borg_archive/prune"} 0
response_time{quantile = "0.99", path = "borg_archive/prune"} 0
response_time{quantile = "0.999", path = "borg_archive/prune"} 0
response_time{quantile = "0.9999", path = "borg_archive/prune"} 0
